src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -0,0 +1,483 @@
+                }
+
+                var results = new MultiDictionary<string, string>();
+
+                // Find matching extension methods from source.
+                foreach (var info in syntaxBuilder)
+                {
+                    // Add simple extension methods with matching target type name
+                    foreach (var targetTypeName in targetTypeNames)
+                    {
+                        var methodInfos = info.SimpleExtensionMethodInfo[targetTypeName];
+                        if (methodInfos.Count == 0)
+                        {
+                            continue;
+                        }
+
+                        foreach (var methodInfo in methodInfos)
+                        {
+                            results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                        }
+                    }
+
+                    // Add all complex extension methods, we will need to completely rely on symbols to match them.
+                    foreach (var methodInfo in info.ComplexExtensionMethodInfo)
+                    {
+                        results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                    }
+                }
+
+                // Find matching extension methods from metadata
+                foreach (var info in symbolBuilder)
+                {
+                    var methodInfos = info.GetMatchingExtensionMethodInfo(targetTypeNames);
+                    foreach (var methodInfo in methodInfos)
+                    {
+                        results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                    }
+                }
+
+                return results;
+            }
+            finally
+            {
+                peReferences.Free();
+            }
+        }
+
+        private static ImmutableArray<SerializableImportCompletionItem> GetExtensionMethodItems(
+            INamespaceSymbol rootNamespaceSymbol,
+            ITypeSymbol receiverTypeSymbol,
+            SemanticModel semanticModel,
+            int position,
+            ISet<string> namespaceFilter,
+            MultiDictionary<string, string> methodNameFilter,
+            StatisticCounter counter,
+            CancellationToken cancellationToken)
+        {
+            var compilation = semanticModel.Compilation;
+            using var _ = ArrayBuilder<SerializableImportCompletionItem>.GetInstance(out var builder);
+
+            using var conflictTypeRootNode = new ConflictNameNode(name: string.Empty);
+
+            foreach (var (fullyQualifiedContainerName, methodNames) in methodNameFilter)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var indexOfLastDot = fullyQualifiedContainerName.LastIndexOf('.');
+                var qualifiedNamespaceName = indexOfLastDot > 0 ? fullyQualifiedContainerName.Substring(0, indexOfLastDot) : string.Empty;
```

Comment:
that's a pity.  so we force a string-alloc on *every* item found?  is this ok because the expetation is that the filter will hav ea small number of results?

Reply:
yes, it's on every containing type we found. And yes, with filter, this number should be fairly small so it's unlikely to cause perf issues.